File: 2021/julia/src/Day2.jl
module Day2

include("Aoc.jl")
using .Aoc

tr = Dict("forward" => [1,0], "up" => [0,-1], "down" => [0,1])

instructions = Aoc.input_lines(2)|>
    lines -> map(l -> split(l, ' '), lines)|>
    lines -> map(s -> tr[s[1]] .* parse(Int64, s[2]), lines)

part1() = prod(sum(instructions))

function part2()
    aim, pos = [0,0], [0,0]
    for i in instructions
        i[1] === 0 ? aim += i : pos += aim * i[1] + i
    end
    prod(pos)
end

end

File: 2021/typescript/src/days/day2.ts
import {Day, Vector} from "../aoc";

export class Day2 extends Day {
    day = (): number => 2;

    private course: Vector[];

    part1 = () => {
        const pos = this.course.reduce((pos, inst) => pos.add(inst), Vector.zero());
        return (pos.x * pos.y);
    }

    part2 = () => {
        let [pos, aim] = [Vector.zero(), Vector.zero()];
        this.course.forEach(inst => inst.y !== 0 ? aim = aim.add(inst) : pos = pos.add(aim.times(inst.x)).add(inst));
        return (pos.x * pos.y);
    }

    setup = () => {
        const dirs = { 'forward': Vector.east(), 'down': Vector.south(), 'up': Vector.north() };
        this.course = this.input.map(l => l.split(' ')).map(s => dirs[s[0]].times(Number(s[1])));
    };
}